Add basic Node API

mattwebbio opened the pull request:
## Description

Hi @msrose 👋 

I've spent some time poking through this project and I really like where you're headed with this! I'd like to use some of the underlying logic (namely `run` and `loadAdapter`, of course) in a separate CLI tool I'm building at @appfolio. I've forked your repo, so if you're hoping to abstract the API further (instead of just exporting these two methods), I can continue using this as is, but thought I'd at least open a PR for discussion.

## Testing

Not much has to be done to test this as its just exporting other code. I've tested these exports locally and they work fine, of course.

## Checklist

I have:
* [ ] ~~Added new automated tests for any new functionality~~ _No new functionality_
* [x] Run `npm run build`

## Licensing statement (do not modify)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

msrose commented:
Hey @mattwebbio thanks for this! Glad to hear you're finding the project useful so far. Could you please elaborate on your use case? I'm interesting to learn more about what you're building since I think it would help inform how we evolve universal-test-runner as we work towards 1.0.0.

I think your change makes sense as an initial Node.js API -- I'll just ask that we move the exported ProtocolResult type to the types package. I'll leave a comment on the diff with further explanation.

I'll also note here that any Node.js API will remain "unstable" for the near future, since our priority for 1.0 release is the CLI. I don't expect any massive breaking changes to happen very quickly, and of course we will document anything that does change in release notes, but please do be prepared for possible breaking changes in minor versions as we flesh out the CLI functionality. Let me know if you have any questions or concerns on this point :)

msrose commented on packages/universal-test-runner/src/index.ts:
```diff
@@ -1,4 +1,6 @@
-// TODO: Node.js API
+export { run } from '../src/run'
+export { loadAdapter, builtInAdapters } from '../src/loadAdapter'
+export { readProtocol, ProtocolResult } from '../src/readProtocol'
```
As mentioned in the main thread, can we please move ProtocolResult to the `@aws/universal-test-runner-types` package? You'll be able to depend on that one as well as `@aws/universal-test-runner` -- even though it's currently marked in the README as being a "private" package, third-party adapter authors will eventually have to depend on it, so we plan on keeping that package mostly as-is.

mattwebbio commented:
> Hey @mattwebbio thanks for this! Glad to hear you're finding the project useful so far. Could you please elaborate on your use case? I'm interesting to learn more about what you're building since I think it would help inform how we evolve universal-test-runner as we work towards 1.0.0.

Certainly! I'm on a developer productivity team building internal tooling. While most of our codebase is in Ruby on Rails and React, we have a variety of internal codebases that use a variety of test frameworks. I'm building a 'test' plugin for our internal CLI tool, making it easier to run tests in repos without a common test setup. This will probably look like a simple 'test' command that wraps `universal-test-runner` but adds some logic to automagically detect the test adapter necessary for the current service.

As a result, this means that I'll be writing test adapters for Ruby's minitest/rspec, and JS's mocha.

> I think your change makes sense as an initial Node.js API -- I'll just ask that we move the exported ProtocolResult type to the types package. I'll leave a comment on the diff with further explanation.
> 
> I'll also note here that any Node.js API will remain "unstable" for the near future, since our priority for 1.0 release is the CLI. I don't expect any massive breaking changes to happen very quickly, and of course we will document anything that does change in release notes, but please do be prepared for possible breaking changes in minor versions as we flesh out the CLI functionality. Let me know if you have any questions or concerns on this point :)

All makes perfect sense. I've updated this as you requested, and I've added another commit to include type definitions in the main package. Does this look good to you?

msrose commented:
> This will probably look like a simple 'test' command that wraps universal-test-runner but adds some logic to automagically detect the test adapter necessary for the current service.

Awesome! We have actually played around with framework "inference" a bit with universal-test-runner, but put it to the side to minimize up-front complexity -- would love to see what you come up with, so please open-source anything you can :)

> As a result, this means that I'll be writing test adapters for Ruby's minitest/rspec, and JS's mocha.

Excellent, looking forward to it! Please let us know if you have any questions/suggestions for writing adapters as you work your way through.

msrose commented:
Changes LGTM, I'll trigger the build again and then squash & merge, and aim to publish a new version by EOD tomorrow.

mattwebbio commented:
Sorry, should have tested locally. Pushing up a fix...

mattwebbio commented:
Done. Builds locally now 😄

msrose commented:
No worries, that's why we have CI 😄

msrose commented:
0.3.0 published just now with the Node.js API available
